File: array/easy/best_time_to_buy_and_sell_stock_121.go
package easy

// link https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

/*
Python solution

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        left, right, maxProfit = 0, 1, 0

        while right < len(prices):
            while right < len(prices) and prices[right] - prices[left] > 0:
                if maxProfit < prices[right] - prices[left]:
                    maxProfit = prices[right] - prices[left]
                right = right + 1
            left = right
            right = left + 1
        return maxProfit
*/

func MaxProfit(prices []int) (maxProfit int) {
	left, right := 0, 1

	for right < len(prices) {
		for right < len(prices) && prices[right]-prices[left] > 0 {
			if maxProfit < prices[right]-prices[left] {
				maxProfit = prices[right] - prices[left]
			}
			right++
		}
		left = right
		right = left + 1
	}
	return maxProfit
}

File: array/easy/java_solutions/BestTimeToBuyAndSellStock_121.java
package array.easy.java_solutions;

import java.util.Arrays;

public class BestTimeToBuyAndSellStock_121 {

    public int maxProfit(int[] prices) {
        int left = 0, right = 1, maxProfit = 0;

        while (right < prices.length) {
            while (right < prices.length && prices[right] - prices[left] > 0) {
                if (maxProfit < prices[right] - prices[left]) {
                    maxProfit = prices[right] - prices[left];
                }
                right++;
            }
            left = right;
            right = left + 1;
        }
        return maxProfit;
    }

    public int maxProfitDivideAndConquer(int[] prices) {
      return -1;
    }

    public static void main(String[] args) {
        int i = new BestTimeToBuyAndSellStock_121().maxProfitDivideAndConquer(new int[]{7, 1, 5, 3, 6, 4});
        System.out.println(i);
    }
}
